contracts/ButtonTokenWamplRouter.sol

```diff
@@ -0,0 +1,105 @@
+     *  Returns the output buttonTokens to the user.
+     * @dev Validates amount > 0
+     * @dev Validates buttonToken has WAMPL as underlying to prevent asset loss
+     * @dev buttonToken is kept as parameter to support multiple buttonToken implementations
+     *
+     * @param buttonToken the button token to deposit into
+     * @param amplAmount The amount of AMPL being deposited
+     * @return The amount of ButtonTokens created
+     */
+    function wamplWrapAndDeposit(IButtonToken buttonToken, uint256 amplAmount)
+        external
+        nonReentrant
+        returns (uint256)
+    {
+        if (amplAmount == 0) revert ZeroAmount();
+        if (buttonToken.underlying() != address(wampl)) revert InvalidButtonAsset();
+        // Transfer ampl to router
+        IERC20(wampl.underlying()).transferFrom(msg.sender, address(this), amplAmount);
+        // Approve ampl from router to wampl
+        IERC20(wampl.underlying()).safeApprove(address(wampl), amplAmount);
+        // Depositing ampl for wampl
+        uint256 wamplAmount = wampl.deposit(amplAmount);
+        // Approving wampl to buttonToken contract
+        IERC20(wampl).safeApprove(address(buttonToken), wamplAmount);
+        // Depositing wampl for buttonToken sent to user
+        return buttonToken.depositFor(msg.sender, wamplAmount);
+    }
+
+    /**
+     * @notice Withdraw the given amount of button tokens from the given ButtonToken and convert to AMPL
+     *  Returns the output AMPL to the user
+     *
+     * @param buttonToken the button token to burn from
+     * @param amount The amount of ButtonTokens to burn
+     * @return The amount of ampl tokens returned
+     */
+    function wamplBurnAndUnwrap(IButtonToken buttonToken, uint256 amount)
+        external
+        nonReentrant
+        returns (uint256)
+    {
+        // Transfer buttonToken to router
+        buttonToken.transferFrom(msg.sender, address(this), amount);
+        // Burn buttonToken to wampl
+        buttonToken.burn(amount);
+        // Burn wampl to ampl
+        uint256 amplAmount = wampl.burn(wampl.balanceOf(address(this)));
+        // Transfer ampl to user
+        IERC20(wampl.underlying()).safeTransfer(msg.sender, amplAmount);
+        return amplAmount;
+    }
+
+    /**
+     * @dev Withdraw all button tokens from the given ButtonToken and convert to AMPL
+     *  Returns the output AMPL to the user
+     *
+     * @param buttonToken the button token to burn from
+     * @return The amount of ampl tokens returned
+     */
+    function wamplBurnAndUnwrapAll(IButtonToken buttonToken)
+        external
+        nonReentrant
+        returns (uint256)
+    {
+        // Transfer all buttonToken to router
+        buttonToken.transferAllFrom(msg.sender, address(this));
+        // Burn all buttonToken to wampl
+        IButtonWrapper(buttonToken).burnAll();
+        // Burn wampl to ampl
+        uint256 amplAmount = wampl.burn(wampl.balanceOf(address(this)));
```

Comment:
Similar to before, we can use burnAll return value

Reply:
Alternatively we can use wampl's burnAll method
https://github.com/ampleforth/ampleforth-contracts/blob/master/contracts/WAMPL.sol#L110